source/Nevermore/Util/DataModificationQueryBuilder.cs

```diff
@@ -330,87 +329,160 @@ CommandParameterValues GetDocumentParameters(Func<DocumentMap, string> allocateI
+                        {
+                            commands.Add(new PreparedCommand(insertStatementBuilder.ToString(), parameters, RetriableOperation.Insert, mapping, options.CommandTimeout));
+                            insertStatementBuilder = new StringBuilder();
+                            parameters = new CommandParameterValues();
+                            parameters.AddRange(documentParametersCreator());
+                        }
+                    }
                 }
             }
+
+            commands.Add(new PreparedCommand(insertStatementBuilder.ToString(), parameters, RetriableOperation.Insert, mapping, options.CommandTimeout));
+            return commands.ToArray();
         }
 
-        string AppendRelatedDocumentStatementsForUpdate(
-            string statement,
-            CommandParameterValues parameters,
+        PreparedCommand[] CreateUpdateCommandsWithRelatedDocuments(
+            string updateStatement,
             DocumentMap mapping,
-            object document)
+            object document,
+            UpdateOptions updateOptions)
         {
+            Func<CommandParameterValues> parametersCreator = () => GetDocumentParameters(
+                m => throw new Exception("Cannot update a document if it does not have an ID"),
+                null,
+                null,
+                document,
+                mapping,
+                DataModification.Update
+            );
+            var commands = new List<PreparedCommand>();
             var relatedDocumentData = GetRelatedDocumentTableData(mapping, new[] {document});
+            var parameters = parametersCreator();
+
             if (relatedDocumentData.Count == 0)
-                return statement;
+            {
+                commands.Add(new PreparedCommand(updateStatement, parameters, RetriableOperation.Update, mapping, updateOptions.CommandTimeout));
+                return commands.ToArray();
+            }
 
             var sb = new StringBuilder();
-            sb.AppendLine(statement);
+            sb.AppendLine(updateStatement);
             sb.AppendLine();
 
-            if (relatedDocumentData.Any(d => d.Related.Any()))
-                sb.AppendLine("DECLARE @references as TABLE (Reference nvarchar(400), ReferenceTable nvarchar(400))");
+            string tableVariableName = "@references";
+            if (parameters.Count + relatedDocumentData.Sum(x => x.Related.Length) > SqlServerParameterLimit) // we need multiple sql commands
+            {
+                // sql commands are executed inside sp_executesql which means table variables and local temp tables won't live between commands
+                tableVariableName = $"##{configuration.RelatedDocumentsGlobalTempTableNameGenerator()}";
+                sb.AppendLine($"CREATE TABLE {tableVariableName} (Reference nvarchar(400) COLLATE SQL_Latin1_General_CP1_CS_AS, ReferenceTable nvarchar(400) COLLATE SQL_Latin1_General_CP1_CS_AS)").AppendLine();
```

Comment:
Why specify collation here?

Reply:
When I run this locally without I get a collation exception if I don't specify the collation here. Looking at my `tempdb` the collation is `Latin1_General_CI_AS` while my actual databases are all `SQL_Latin1_General_CP1_CI_AS`.

Maybe the collation needs to be configurable in the configuration in case someone else using this happens to have a different collation on their databases?

Reply:
So it looks like the column collation in the temporary table needs to match the collation from the related document table column when performing the join. The reason why `SQL_Latin1_General_CP1_CS_AS` is there is because that's how we set up the test database.

We would either need to:
1. query the database schema to determine what the collation of the related document table columns are, or
2. find a way around not using global temp tables

Reply:
Pushed change with collation defined in configuration

Reply:
Continued https://github.com/OctopusDeploy/Nevermore/pull/153/files#r659409108